renku_notebooks/util/kubernetes_.py

```diff
@@ -63,25 +64,35 @@
     )
 
 
-def get_user_server(user, namespace, project, commit_sha):
-    """Fetch the user named server"""
-    RENKU_ANNOTATION_PREFIX = config.RENKU_ANNOTATION_PREFIX
-    servers = get_user_servers(user)
-    for server in servers.values():
-        annotations = server["annotations"]
-        if (
-            annotations.get(RENKU_ANNOTATION_PREFIX + "namespace") == namespace
-            and annotations.get(RENKU_ANNOTATION_PREFIX + "projectName") == project
-            and annotations.get(RENKU_ANNOTATION_PREFIX + "commit-sha") == commit_sha
-        ):
-            current_app.logger.debug(server)
-            return server
-    return {}
+def get_user_servers(user, namespace=None, project=None, branch=None, commit_sha=None):
+    """Fetch all the user named servers that matches the provided criteria"""
+
+    def filter_server(server):
+        def check_annotation(annotation, value):
+            annotation_name = config.RENKU_ANNOTATION_PREFIX + annotation
+            return not value or server["annotations"].get(annotation_name) == value
+
+        return (
+            check_annotation("namespace", namespace)
+            and check_annotation("projectName", project)
+            and check_annotation("branch", branch)
+            and check_annotation("commit-sha", commit_sha)
+        )
+
+    servers = _get_all_user_servers(user)
+    filtered_servers = [s for s in servers if filter_server(s)]
+    current_app.logger.debug(filtered_servers)
+    return {s["name"]: s for s in filtered_servers}
```

Comment:
why return just the names and not the full server JSON?

Reply:
This returns a map of {server_name: server}. 
I've changed the code here according to your other comment and this is a bit simpler now.

Reply:
argh yes of course. apologies for the stupid question.